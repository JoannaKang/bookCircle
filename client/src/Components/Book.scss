$primary-color: #DA291C;
$primary-dark: #1C191A;
$background-color: #F6E7D8;

.BookDisplayDiv {
  margin: 0 1rem;
  position: relative;

  .book {
    height: 220px;
    width: 150px;
    box-shadow:  5px 5px 5px  #f1dac3;
  }
  .star {
    color: gold;
    height: 2rem;
    width: 2rem;
    position: absolute;
    z-index: 1;
    bottom: 10px;
    left: 2px;
  }
}

.BookDisplayDiv:last-child {
  padding-right: 1rem;
}

.BookDisplayDiv:first-child {
  margin-left: 2.2rem;
}