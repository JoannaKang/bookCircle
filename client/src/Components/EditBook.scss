$primary-color: #DA291C;
$primary-dark: #1C191A;
$background-color: #F6E7D8;


.EditBookPopOutDiv {
  display: flex;
  position: absolute;
  border: 5px solid $primary-dark;
  border-radius: 10px;
  background-color: $background-color;
  max-height: 70vh;
  min-height: 70vh;
  min-width: 90vw;
  max-width: 90vw;
  align-items: center;
  z-index: 2;

  .EditBookH1 { margin-bottom: 2rem;}

  .EditBookPopOutContainer {
    display: flex;

    flex-direction: column;
  }

  .EditOptionContainer {
    display: flex;
    align-items: center;
  }

  h2 {
    margin: 1rem;
  }

  textarea {
    height: 80px;
    width: 50vw;
  }

  select {
    height: 30px;
    margin-left: 1rem;
  }

  .star,
  .notStar,
  .saveChanges,
  .deleteBtn {
    margin: 0.5rem;
    border: none;
    border-radius: 5px;
    padding: 0.5rem 0rem;
    width: 4rem;
    font-weight: bold;
    font-size: 1rem;
  }
  .starReadBtnsDiv {
    display: flex;
  }
  .star {
    background-color: $primary-color;
    color: $background-color;
  }
  .notStar {
    background-color: $primary-dark;
    color: $background-color;
  }

  .EditActionBtns {
    display: flex;
    justify-content: space-between;
  }

  .saveChanges {
    background-color: green;
    width: 50vw;
  }

  .deleteBtn {
    background-color: red;
  }
}