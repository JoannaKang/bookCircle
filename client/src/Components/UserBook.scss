$primary-color: #DA291C;
$primary-dark: #1C191A;
$background-color: #F6E7D8;

.UserBookClickedDiv {
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  z-index: 10;
  background-color: rgba(182, 182, 182, 0.8);
  display: flex;
  align-items: center;
  justify-content: center;
} 

.UserBookPopOutDiv {
  display: flex;
  position: relative;
  border: 5px solid $primary-dark;
  border-radius: 10px;
  background-color: $background-color;
  max-height: 40vh;
  min-height: 40vh;
  min-width: 90vw;
  max-width: 90vw;
  align-items: center;

}

.SelectedBookDisplayDiv {
  position: relative;
  margin: 0 1rem;
  
  img {
    max-height: 240px;
  }

  .SelectedBookStar {
    color: gold;
    height: 2rem;
    width: 2rem;
    position: absolute;
    z-index: 1;
    bottom: 10px;
    left: 2px;
  }
}

.UserBookPopOutContentContainer {
  min-height: 400px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-right: 0.5rem;

  h3 {
    font-weight: 300;
    font-size: 1.4rem;
    margin: 1rem 0 0 0;
  }

  h4 {
    font-weight: 300;
    font-size: 1.3rem;
    margin: 1rem 0 0 0;
  }
  p {
    margin: 0;
  }

  input {
    margin-bottom: 1rem;
    overflow: none;
  }
}

.editBookBtn {
  background-color: $primary-dark;
  color: $background-color;
  font-size: 1.2rem;
  border-radius: 10px;
  border:none;
  padding: 0.5rem;
  width: 50px;
  margin-top: 1rem;
}

.UserBookEscapeButton {
  position: absolute;
  top: 0;
  right: 0;
  width: 30px;
  height: 30px;
}

