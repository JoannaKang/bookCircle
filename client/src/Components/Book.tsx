import React, { FunctionComponent } from 'react';
import { Book, BorrowableBook } from '../Interfaces';
import { AiFillStar } from "react-icons/ai";
import './Book.scss'


type BookProps = {
  book: any,
  handleBookClicked: Function,
}

const BookDisplay: FunctionComponent<BookProps> = ({book, handleBookClicked}) => {

  const handleClick = () => {
    handleBookClicked(book)
  }

  return (
    <div className="BookDisplay">
      <img className="book" 
      src={book.imageUrl} 
      alt={`${book.title} book cover`}
      onClick={handleClick}
      />
      {book.star && <AiFillStar className="star" />}
    </div>
  )
}

export default BookDisplay;
