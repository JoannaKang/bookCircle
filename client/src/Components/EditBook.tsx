import React, { FunctionComponent, useEffect, useState } from "react";
import { Book } from "../Interfaces";
import { AiFillCloseCircle } from "react-icons/ai";

import './EditBook.scss'

type EditBookProps = {
  book: Book,
  deleteFunc: Function,
  update: Function,
  setEditBook: Function,
  userId: string | null,
  setClickedBook: Function
}

const EditBook: FunctionComponent<EditBookProps> = ({deleteFunc, setEditBook, book, userId, setClickedBook}) => {
  const [updatedBook, setUpdatedBook] = useState<any>(
    {
    title: undefined,
    authors: undefined,
    imageUrl: undefined,
    dateRead: undefined,
    review: undefined,
    availableToBorrow: undefined,
    genre: undefined,
    star: undefined,
  }
  )

  useEffect(() => {
    setUpdatedBook(book)
    // return () => setUpdatedBook(undefined)
  },[])

  const handleCloseClick = () => {
    setEditBook(false)
  }

  const handleReviewChange = (e:React.ChangeEvent<HTMLTextAreaElement> ) => {
    setUpdatedBook(
      {
      ...updatedBook,
      review: e.target.value
    })
  }

  const handleGenreChange = (e:React.ChangeEvent<HTMLSelectElement> ) => {
    setUpdatedBook(
      {
      ...updatedBook,
      genre: e.target.value
    })
  }

  const handleStarChange = () => {
    setUpdatedBook(
      {
        ...updatedBook,
        star: !(updatedBook.star)
      })
  }

  const handleAvailabilityChange = () => {
    setUpdatedBook(
      {
        ...updatedBook,
        availableToBorrow: !(updatedBook.availableToBorrow)
      })
  }

  const handleDateChange = (e:React.ChangeEvent<HTMLInputElement> ) => {
    setUpdatedBook(
      {
      ...updatedBook,
      dateRead: e.target.value
    })
  }

  const handleClick = () => {
    console.log('UPDATED',updatedBook)
  }

  const handleDelete = async () => {
    await deleteFunc(userId,book.id)
    setUpdatedBook({
      title: undefined,
      authors: undefined,
      imageUrl: undefined,
      dateRead: undefined,
      review: undefined,
      availableToBorrow: undefined,
      genre: undefined,
      star: undefined,
    })
    setEditBook(false)
    setClickedBook(undefined)
  }

  console.log('BOOK', book)
  return(
    <div className="EditBookPopOutDiv">
      {book && <div className="EditBookPopOutContainer">
        <h1 className="EditBookH1">Update Book:</h1>
        <AiFillCloseCircle onClick={handleCloseClick} className="UserBookEscapeButton"/>
        <div className="EditOptionContainer">
          <h2>Review:</h2>
          <textarea 
            className="editookReviewInput" 
            maxLength={100} 
            value={updatedBook ? updatedBook.review: ''} 
            onChange={handleReviewChange}  
          />
        </div>
        <div className="EditOptionContainer">
          <h2>Genre:</h2>
          <select value={updatedBook ? updatedBook.genre: 'fiction'} id="pickGenre" onChange={handleGenreChange} >
            <option value="fiction">Fiction</option>
            <option value="non-fiction">Non-Fiction</option>
            <option value="crime">Crime & Thriller</option>
            <option value="science fiction">Science Fiction</option>
            <option value="fantasy">Fantasy</option>
            <option value="action & adventure">Action & Adventure</option>
            <option value="politics & history">Politics & History</option>
            <option value="romance">Romance</option>
            <option value="comedy">Comedy</option>
            <option value="science & technology">Science & Technology</option>
            <option value="biography">Biography</option>
            <option value="arts & culture">Arts & Culture</option>
            <option value="self-improvement">Self-Improvement</option>
            </select>
        </div>
        <div className="EditOptionContainer">
          <h2>Star read:</h2>
          <div className="starReadBtnsDiv">
            <button className={updatedBook.star ? 'star' : 'notStar'} onClick={handleStarChange}>yes</button>
            <button className={updatedBook.star ? 'notStar' : 'star'} onClick={handleStarChange}>no</button>
          </div>
        </div>
        <div className="EditOptionContainer">
          <h2>Can lend:</h2>
          <div className="starReadBtnsDiv">
            <button className={updatedBook.availableToBorrow ? 'star' : 'notStar'} onClick={handleAvailabilityChange}>yes</button>
            <button className={updatedBook.availableToBorrow ? 'notStar' : 'star'} onClick={handleAvailabilityChange}>no</button>
          </div>
        </div>
        <div className="EditOptionContainer">
          <h2>Date read:</h2>
          <input type="date" value={updatedBook ? updatedBook.dateRead : ''} onChange={handleDateChange}/>
        </div>
        <div className="EditActionBtns">
          <button className="saveChanges" onClick={handleClick}>save</button>
          <button className="deleteBtn" onClick={handleDelete}>delete</button>
        </div>
      </div>}
    </div>
  )
}

export default EditBook;