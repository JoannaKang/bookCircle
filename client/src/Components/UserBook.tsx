import React, { FunctionComponent, useEffect, useState } from "react";
import { Book, User } from '../Interfaces'
import { AiFillStar, AiFillCloseCircle } from "react-icons/ai";
import EditBook from '../Components/EditBook'


import './UserBook.scss'

type UserBookProps = {
  clickedBook: Book,
  handleClosePopOut: Function,
  updateAvailableBooks?: Function,
  user: User,
  removeBookFromBookCase: Function,
  setClickedBook: Function,
  editBookFunc: Function
}
const UserBook: FunctionComponent<UserBookProps> = (
  {
    clickedBook,
    handleClosePopOut,
    user,
    removeBookFromBookCase,
    setClickedBook,
    editBookFunc
  }) => {
  const [editBook, setEditBook] = useState<boolean>(false)

  const handleEditClick = () => {
    console.log('HandleEditClick')
    setEditBook(true)
  }

  const handleCloseClick = (): void => {
    handleClosePopOut()
  }

  //TODO: edit popup should open when user click edit button
  //TODO: user can edit book info when click edit button
  //TODO: user can update review /genre / start read / can lend / date read info when click "save""" button
  //TODO: use can delete book info when click "delete" button
  return (
    <div className="UserBookClickedDiv">
      {editBook &&
        <EditBook
          book={clickedBook}
          deleteFunc={removeBookFromBookCase}
          update={() => { }}
          setEditBook={setEditBook}
          userId={user._id}
          setClickedBook={setClickedBook}
          editBookFunc={editBookFunc}
        />}
      <div className="UserBookPopOutDiv">
        <AiFillCloseCircle onClick={handleCloseClick} className="UserBookEscapeButton" />

        <div className="SelectedBookDisplayDiv">
          <img src={clickedBook.imageUrl} alt="imgUrl" />
          {clickedBook.star && <AiFillStar className="SelectedBookStar" />}
        </div>
        <div className="UserBookPopOutContentContainer">
          <h3>{clickedBook.title}</h3>
          <h4>Your Review:</h4>
          <p>{clickedBook.review}</p>
          <h3>{clickedBook.genre}</h3>
          <p>{`Read ${clickedBook.dateRead}`}</p>
          <p>{clickedBook.availableToBorrow ? `Available to Borrow` : `Unavailable to Borrow`}</p>
          <button className="editBookBtn" onClick={handleEditClick}>edit</button>
        </div>
      </div>
    </div>
  )
}

export default UserBook;
