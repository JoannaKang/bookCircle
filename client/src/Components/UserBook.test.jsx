import UserBook from './UserBook';
import EditBook from '../Components/EditBook'
import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import { shallow } from 'enzyme';
import Book from '../Interfaces'
import React, { useState } from 'react';
import { Simulate } from 'react-dom/test-utils';


const mockBook = () => {
  const mockup = {}
  mockup.title = "Harry potter";
  mockup.authors = ["JK rowling"];
  mockup.dateRead = "2020-11-11";
  mockup.review = "fucking awesome";
  mockup.availableToBorrow = true;
  mockup.genre = "Fiction";
  mockup.star = true;

  return mockup;
}

const mockUser = () => {
  const mockup = {}
  mockup.books = [];
  mockup.friends = [];
  mockup._id = 'asdfjaksdjfkal1';
  mockup.name = 'json';
  mockup.password = 'asdf';
  mockup.activityLog = [{}];
  mockup.yearlyTarget = '20';

  return mockup;
}

describe('Edit book', () => {
  it('matches the snapshot when a book prop is provided', () => {
    const bookInfo = { review: 'asdf', genre: 'fiction' }
    const { container } = render(<EditBook book={bookInfo} />)
    expect(container).toMatchSnapshot()
  })

  test('Edit popup opened correctly', () => {
    const { container } = render(<EditBook />);
    // const { container } = render(<div class='EditBookPopOutDiv' ></div>);
    expect(container.firstChild).toHaveClass('EditBookPopOutDiv');
  })


  //FIXME: How to test click event?????
  it('handleClick is called when edit button is clicked', () => {

    const Userbookprops = render(<UserBook clickedBook={mockBook()} user={mockUser()} />);
    const button = Userbookprops.getByText('edit');
    Simulate.click(button);


    expect(screen.getByTestId('editbook')).toBeDefined();


  })
})

