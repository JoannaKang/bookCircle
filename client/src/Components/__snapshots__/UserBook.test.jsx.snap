// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit book matches the snapshot when a book prop is provided 1`] = `
<div>
  <div
    class="EditBookPopOutDiv"
    data-testid="editbook"
  >
    <div
      class="EditBookPopOutContainer"
    >
      <h1
        class="EditBookH1"
      >
        Update Book:
      </h1>
      <svg
        class="UserBookEscapeButton"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 0 1-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
        />
      </svg>
      <div
        class="EditOptionContainer"
      >
        <h2>
          Review:
        </h2>
        <textarea
          class="editookReviewInput"
          maxlength="100"
        >
          asdf
        </textarea>
      </div>
      <div
        class="EditOptionContainer"
      >
        <h2>
          Genre:
        </h2>
        <select
          id="pickGenre"
        >
          <option
            value="fiction"
          >
            Fiction
          </option>
          <option
            value="non-fiction"
          >
            Non-Fiction
          </option>
          <option
            value="crime"
          >
            Crime & Thriller
          </option>
          <option
            value="science fiction"
          >
            Science Fiction
          </option>
          <option
            value="fantasy"
          >
            Fantasy
          </option>
          <option
            value="action & adventure"
          >
            Action & Adventure
          </option>
          <option
            value="politics & history"
          >
            Politics & History
          </option>
          <option
            value="romance"
          >
            Romance
          </option>
          <option
            value="comedy"
          >
            Comedy
          </option>
          <option
            value="science & technology"
          >
            Science & Technology
          </option>
          <option
            value="biography"
          >
            Biography
          </option>
          <option
            value="arts & culture"
          >
            Arts & Culture
          </option>
          <option
            value="self-improvement"
          >
            Self-Improvement
          </option>
        </select>
      </div>
      <div
        class="EditOptionContainer"
      >
        <h2>
          Star read:
        </h2>
        <div
          class="starReadBtnsDiv"
        >
          <button
            class="notStar"
          >
            yes
          </button>
          <button
            class="star"
          >
            no
          </button>
        </div>
      </div>
      <div
        class="EditOptionContainer"
      >
        <h2>
          Can lend:
        </h2>
        <div
          class="starReadBtnsDiv"
        >
          <button
            class="notStar"
          >
            yes
          </button>
          <button
            class="star"
          >
            no
          </button>
        </div>
      </div>
      <div
        class="EditOptionContainer"
      >
        <h2>
          Date read:
        </h2>
        <input
          type="date"
          value=""
        />
      </div>
      <div
        class="EditActionBtns"
      >
        <button
          class="saveChanges"
        >
          save
        </button>
        <button
          class="deleteBtn"
        >
          delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
